command/agent/cache/lease_cache.go

```diff
@@ -0,0 +1,813 @@
+		}, nil
+	}
+
+	c.logger.Debug("forwarding request", "path", req.Request.URL.Path, "method", req.Request.Method)
+
+	// Pass the request down and get a response
+	resp, err := c.proxier.Send(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the namespace from the request header
+	namespace := req.Request.Header.Get(consts.NamespaceHeaderName)
+	// We need to populate an empty value since go-memdb will skip over indexes
+	// that contain empty values.
+	if namespace == "" {
+		namespace = "root/"
+	}
+
+	// Build the index to cache based on the response received
+	index = &cachememdb.Index{
+		ID:          id,
+		Namespace:   namespace,
+		RequestPath: req.Request.URL.Path,
+	}
+
+	secret, err := api.ParseSecret(bytes.NewBuffer(resp.ResponseBody))
+	if err != nil {
+		c.logger.Error("failed to parse response as secret", "error", err)
+		return nil, err
+	}
+
+	isRevocation, err := c.handleRevocationRequest(ctx, req, resp)
+	if err != nil {
+		c.logger.Error("failed to process the response", "error", err)
+		return nil, err
+	}
+
+	// If this is a revocation request, do not go through cache logic.
+	if isRevocation {
+		return resp, nil
+	}
+
+	// Fast path for responses with no secrets
+	if secret == nil {
+		c.logger.Debug("pass-through response; no secret in response", "path", req.Request.URL.Path, "method", req.Request.Method)
+		return resp, nil
+	}
+
+	// Short-circuit if the secret is not renewable
+	tokenRenewable, err := secret.TokenIsRenewable()
+	if err != nil {
+		c.logger.Error("failed to parse renewable param", "error", err)
+		return nil, err
+	}
+	if !secret.Renewable && !tokenRenewable {
+		c.logger.Debug("pass-through response; secret not renewable", "path", req.Request.URL.Path, "method", req.Request.Method)
+		return resp, nil
+	}
+
+	var renewCtxInfo *ContextInfo
+	switch {
+	case secret.LeaseID != "":
+		c.logger.Debug("processing lease response", "path", req.Request.URL.Path, "method", req.Request.Method)
+		entry, err := c.db.Get(cachememdb.IndexNameToken, req.Token)
+		if err != nil {
+			return nil, err
+		}
+		// If the lease belongs to a token that is not managed by the agent,
+		// return the response without caching it.
+		if entry == nil {
+			c.logger.Debug("pass-through lease response; token not managed by agent", "path", req.Request.URL.Path, "method", req.Request.Method)
+			return resp, nil
+		}
+
+		// Derive a context for renewal using the token's context
+		newCtxInfo := new(ContextInfo)
+		newCtxInfo.Ctx, newCtxInfo.CancelFunc = context.WithCancel(entry.RenewCtxInfo.Ctx)
+		newCtxInfo.DoneCh = make(chan struct{})
+		renewCtxInfo = newCtxInfo
+
+		index.Lease = secret.LeaseID
+		index.LeaseToken = req.Token
+
+	case secret.Auth != nil:
+		c.logger.Debug("processing auth response", "path", req.Request.URL.Path, "method", req.Request.Method)
+		isNonOrphanNewToken := strings.HasPrefix(req.Request.URL.Path, vaultPathTokenCreate) && resp.Response.StatusCode == http.StatusOK && !secret.Auth.Orphan
```

Comment:
isNonOrphan will also be true when creating a token via the create-orphan endpoint or any role.

Reply:
`isNonOrphanNewToken` will not be true when a token is created via create-orphan. Note that we are looking for non-orphan case here, not orphan cases. Orphan tokens will get cached. For non-orphan tokens, we want the renewer context to be derived from its parent token. The check seems fine unless I am missing something.

Reply:
Ah, no. I'm missing something -- scrolling to the right. When I initially looked at this it cut off at just the right spot for me to not actually see the last part of the boolean expression.

That said, you should make that the first part -- both because it's a much faster check and it's the key part of the check.

Also -- why are the other parts of the expression necessary? Why not just gate purely on the Orphan check?

Reply:
You are right. Other parts of the expression is unnecessary. Calvin will be removing it in https://github.com/hashicorp/vault/pull/6322.